Flow policy experimental docs (#12523)
RasaHQ/rasa: docs/docs/llms/flow-policy.mdx

sidebar_label: Flow Policy
title: Flow Policy
abstract: |
  Guide on implementing and managing Flows with the FlowPolicy.
---

import RasaDiscoveryBanner from "@theme/RasaDiscoveryBanner";
import flowStackEvolution from "./flow_stack_evolution.png";
import flowStackOverview from "./flow_stack.png";

<RasaDiscoveryBanner />

The Flow Policy in Rasa is a state machine that enables you to
manage conversational flows in your chatbot and combine your business logic 
with machine learning. 

The policy manages the state of your bot, transitioning between states, and

> initiating new flows when required. The "Flow Stack" represents 

Reviewer: Do we want to explain that the flow stack is LIFO? I feel like most developers know what a stack is, but not assuming prior knowledge is one of our documentation principles. Maybe something for later though. Up to you

Author: will do so further down :+1:

all flows that have been started and not yet completed and reference steps
defined using the [flows specification](../flows.mdx).

<Image
  img={flowStackOverview}
  caption="Flow Stack reflecting the state of the conversation referencing a flow."
  alt=""
/>

The policy allows you to express your logic in flows, shaping the way a conversation 
progresses based on user input and system response. Other classifiers will
modify the underlying state machine to handle edge-cases. 

Please refer to the [flows specification](../flows.mdx) for detailed 
information on defining flows.

## Adding the FlowPolicy to your bot

Incorporating the FlowPolicy into your bot involves a two-step process. First,

> you need to configure the FlowPolicy in your 'config.yml' file. The

Reviewer: ```suggestion
you need to configure the `FlowPolicy` in your `config.yml` file. The
```

states, and initiating new flows when required.

### Managing the State

The state of your bot is managed using a 'flow stack'. The flow stack stores the
states of the conversation, enabling the bot to keep track of where in the flow
it currently is. The flow stack is stored in the `flow_stack` slot.

Let's take the following example flow and a conversation that is currently
using it:
```yaml title="flows.yml"
flows:
  transfer_money:
    description: This flow let's users send money to friends and family.

    steps:
      - id: "start"
        intent: transfer_money
        next: "ask_recipient"
      - id: "ask_recipient"
        question: transfer_money_recipient
        next: "ask_amount"
      - id: "ask_amount"
        question: transfer_money_amount_of_money
        next: "check_transfer_funds"
    # - ...
```

<Image
  img={flowStackEvolution}
  caption="Changes to the flow stack during a correction of a prior user input"
  alt=""
/>

The correction *"oh sorry I meant John"* from the user triggers a correction
flow to be started. This **correction is not explicitly modeled** in the 
`transfer_money` flow but rather handled by a so called 
"conversational pattern".

> The correction flow is one of these "conversational patterns". It is only 
> executed when the user corrects a prior input. When triggered, the 
> correction flow is put on top of the flow stack and executed. 

Reviewer: ```suggestion
The correction flow is only  executed when the user wants to correct a prior input.
When triggered, the correction flow is put on top of the flow stack and executed. 
```

The correction flow is an "interruption" to the `transfer_money` flow. Once the

> correction flow is completed, the `transfer_money` flow is resumed. Before, the

Reviewer: ```suggestion
correction flow is completed, the `transfer_money` flow is resumed. Before the
```

`transfer_money` is resumed, the `resume_flow` flow is triggered. It allows the

> dialogue designer to define how to guide the user back to the original flow, 

Reviewer: I'm not sure how it technically allows the dialogue designer to define how to guide the user back to the orginal flow. Can you detail?

`transfer_money` in this case.

Once the `resume_flow` flow is completed, the `transfer_money` flow is resumed
and the bot is back in the state it was before the correction flow was started.
The `transfer_money` flow will re-ask the question that was interrupted by the
correction flow.

### Transitioning Between States

The FlowPolicy transitions between states in response to user inputs or after an
action has been completed. Initially, it waits for user input. Once received, it
advances the state of the machine until it makes an action prediction. After the
action has been completed and its result incorporated, it continues advancing to
the next state.

### Starting New Flows

> Starting a new flow in the FlowPolicy can be done in one of two ways. Either,

Reviewer: ```suggestion
Starting a new flow in the `FlowPolicy` can be done in one of two ways. Either,
```

by user input that corresponds that matches the first step in a flow or 
by predicting the "flow action".

#### User Input
The user message *"i want to transfer money"* will trigger a flow 
`transfer_money` if this flow is defined to be triggered by this message, e.g.:

```yaml title="flows.yml"

flows:
  transfer_money:
    name: Transfer money
    steps:
      - id: "1"
        intent: transfer_money
        action: action_transfer_money
      - id: "2"
        action: utter_lets_transfer_money
    # - ... 
```

The NLU model will predict the intent `transfer_money` and the FlowPolicy will
start the flow `transfer_money` as a result.

#### Predicting Flow Actions
Another policy can predict the action `flow_transfer_money` and the FlowPolicy

> will start the flow `transfer_money` as a result.

Author: added an faq :+1:
